code/xcvm/cosmwasm/contracts/router/src/contract.rs

```diff
@@ -0,0 +1,368 @@
+use crate::{
+	error::ContractError,
+	msg::{ExecuteMsg, InstantiateMsg, QueryMsg},
+	state::{Config, UserId, CONFIG, INTERPRETERS},
+};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+	from_binary, to_binary, wasm_execute, Addr, Binary, CosmosMsg, Deps, DepsMut, Env, MessageInfo,
+	Reply, Response, StdError, StdResult, SubMsg, WasmMsg, WasmQuery,
+};
+use cw20::{Cw20Contract, Cw20ExecuteMsg};
+use xcvm_asset_registry::msg::{GetAssetContractResponse, QueryMsg as AssetRegistryQueryMsg};
+use xcvm_core::{Displayed, Funds, NetworkId};
+use xcvm_interpreter::msg::{
+	ExecuteMsg as InterpreterExecuteMsg, InstantiateMsg as InterpreterInstantiateMsg,
+};
+
+const INSTANTIATE_REPLY_ID: u64 = 1;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+	deps: DepsMut,
+	_env: Env,
+	_info: MessageInfo,
+	msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+	let addr = deps.api.addr_validate(&msg.registry_address)?;
+	CONFIG.save(
+		deps.storage,
+		&Config { registry_address: addr, interpreter_code_id: msg.interpreter_code_id },
+	)?;
+	Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+	deps: DepsMut,
+	env: Env,
+	_info: MessageInfo,
+	msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+	match msg {
+		ExecuteMsg::Run { network_id, user_id, interpreter_execute_msg, funds } =>
+			handle_run(deps, env, network_id, user_id, interpreter_execute_msg, funds),
+	}
+}
+
+pub fn handle_run(
+	deps: DepsMut,
+	env: Env,
+	network_id: NetworkId,
+	user_id: UserId,
+	interpreter_execute_msg: InterpreterExecuteMsg,
+	funds: Funds<Displayed<u128>>,
+) -> Result<Response, ContractError> {
+	match INTERPRETERS.load(deps.storage, (network_id.0, user_id.clone())) {
+		Ok(interpreter_address) => {
+			let response =
+				send_funds_to_interpreter(deps.as_ref(), interpreter_address.clone(), funds)?;
+			let wasm_msg = wasm_execute(interpreter_address, &interpreter_execute_msg, vec![])?;
+			Ok(response.add_message(wasm_msg))
+		},
+		Err(_) => {
+			let Config { registry_address, interpreter_code_id } = CONFIG.load(deps.storage)?;
+			let instantiate_msg: CosmosMsg = WasmMsg::Instantiate {
+				admin: Some(env.contract.address.clone().into_string()),
+				code_id: interpreter_code_id,
+				msg: to_binary(&InterpreterInstantiateMsg {
+					registry_address: registry_address.into_string(),
+					network_id,
+					user_id: user_id.clone(),
+				})?,
+				funds: vec![],
+				label: format!("xcvm-interpreter-{}-{}", network_id.0, hex::encode(&user_id)),
```

Comment:
may be spec should be updated. we have 4 specs now already. and yet there is on team. not in this PR and may be not you. just mention. will not bother with this further.

Reply:
This is a cosmwasm-specific way of handling internal contracts which is very specific for the general spec. We could state this in somewhere if anyone needs it but not sure they will.